File: 05_排序/10-排序5 PAT Judge.py
N, K, M = map(int, input().split())
full_mark_l = list(map(int, input().split()))
user_info = {}
rank_l = []

if __name__ == "__main__":
    for _ in range(M):
        id_, p, s = map(int, input().split())
        p -= 1
        if id_ not in user_info:
            user_info[id_] = [None for _ in range(K)]
        if user_info[id_][p] is None:
            user_info[id_][p] = s
        elif user_info[id_][p] < s:
            user_info[id_][p] = s
    
    for id_ in user_info.keys():
        score_l = user_info[id_].copy()
        perfect_p_nums = 0
        for i in range(K):
            if score_l[i] == full_mark_l[i]:
                perfect_p_nums += 1
        
        score_l = list(filter(lambda x: x is not None, score_l)) # 去掉score_l中的None
        if all(i == -1 for i in  score_l):                       # 没有一个提交通过编译的人不进入排名
            continue
        
        score_l = [(0 if i == -1 else i) for i in score_l]        # 把未通过编译的-1换成0,表示这题得0分
        rank_l.append((-sum(score_l), -perfect_p_nums, id_))

    rank_l.sort()
    
    rank = 1
    last_score = -rank_l[0][0]
    same_rank_nums = 0
    for line in rank_l:
        score, id_ = -line[0], line[2]
        if score == last_score:
            print(rank, f"{id_:05d}", score, ' '.join(map(str, user_info[id_])).replace("None", '-').replace('-1', '0'))
            same_rank_nums += 1
        else:
            rank += same_rank_nums
            print(rank, f"{id_:05d}", score, ' '.join(map(str, user_info[id_])).replace("None", '-').replace('-1', '0'))
            same_rank_nums = 1
        last_score = score


'''
4 最大随机测试，有满分的题目重复提交 运行超时 0 -- 5028 KB
'''

File: 05_排序/10-排序5 PAT Judge.cpp
#include<cstdio>
#include<algorithm>
#include<vector>
using namespace std;

struct Stu{
    int id, all, pef, r;
    int p[6] = {-2, -2, -2, -2, -2, -2};
} a[10001];

int n, m, k;
int pf[6];

int sum(Stu a){
    int i, s=0;
    
    for (i=1; i<=k; i++){
        if (a.p[i] > 0) s += a.p[i];
    }
    
    return s;
}

int count_pef(Stu a){
    int i, s = 0;
    for (i=1; i<=k; i++){
        if (a.p[i] == pf[i]) s++;
    }
    return s;
}

bool cmp(Stu a, Stu b){
    if (a.all != b.all) return a.all > b.all;
    else if (a.pef != b.pef) return a.pef > b.pef;
    else return a.id < b.id;
}

int main(){
    int i, j;
    scanf("%d %d %d", &n, &k, &m);
    for (i=1; i<=k; i++) scanf("%d ", &pf[i]);
    
    for (i=0; i<m; i++){
        int id, pr, sc;
        scanf("%d %d %d", &id, &pr, &sc);
        a[id].id = id;
        if (a[id].p[pr] < sc){
            a[id].p[pr] = sc;
        }
    }
    
    vector<Stu> rank;
    for (i=1; i<=n; i++){
        int join = 0;
        for (j=1; j<=k; j++){
            if (a[i].p[j] >= 0) {join = 1; break;}
        }
        
        if (join){
            a[i].all = sum(a[i]);
            a[i].pef = count_pef(a[i]);
            rank.push_back(a[i]);
        }
    }
    
    sort(rank.begin(), rank.end(), cmp);
    
    Stu h;
    h.all = -1;
    rank.insert(rank.begin(), h);
    for (i=1; i<(int)rank.size(); i++){
        if (rank[i].all != rank[i-1].all) rank[i].r = i;
        else rank[i].r = rank[i-1].r;
        
        printf("%d %05d %d", rank[i].r, rank[i].id, rank[i].all);
        for (j=1; j<=k; j++){
            if (rank[i].p[j] >= 0) printf(" %d", rank[i].p[j]);
            else if (rank[i].p[j] == -1) printf(" 0");
            else printf(" -");
        }
        printf("\n");
    }
    
    return 0;
}